### PR TITLE
Add How-to & references pages, plus supporting metadata

### PR DESCRIPTION
**Pull request by jmacdotorg**
Update various Hugo-related files in preparation for a subsequent documentation content update.
    
* Add `.hugo_build.lock` to `.gitignore`.
* Update hugo.yaml to allow HTML elements within Markdown source files.
* Add a custom shortcode for wrapping placeholders in code snippets in `<var>` tags.
* Add two shortcodes to store the URL paths to the new and legacy proxy documentation.
* Add custom CSS for putting a non-copyable `$ ` prompt before console-command illustrations.

Add and update the content in `proxy2` with new how-to and reference documentation for installing, configuring, and running the proxy using the standard route library.

### CODE REVIEW

File: themes/hextra/assets/css/custom.css
Comment (dormando) [inline themes/hextra/assets/css/custom.css]: note for maybe just myself: did hugo have a local theme override directory? if we avoid changing files in the themes/hextra directory I can more easily pull updates to the theme in the future.

File: content/features/proxy2/run.md
Comment (toru) [inline content/features/proxy2/run.md]: There's a repetition: `to connect to connect`

File: content/features/proxy2/run.md
Comment (toru) [inline content/features/proxy2/run.md]: "tress"

Comment (jmacdotorg) [reply content/features/proxy2/run.md]: Fixed.

File: content/features/proxy2/run.md
Comment (jmacdotorg) [reply content/features/proxy2/run.md]: Fixed.

File: content/features/proxy2/quickstart.md
Comment (dormando) [inline content/features/proxy2/quickstart.md:1]: should this quickstart be the index page? ie the `_index.md` file in this directory.

File: content/features/proxy2/install.md
Comment (dormando) [inline content/features/proxy2/install.md]: should be `--enable-proxy`

File: content/features/proxy2/install.md
Comment (dormando) [inline content/features/proxy2/install.md]: Should probably double check this, but the when referring to options to `configure` people typically refer to them as `configure options`, or more simply compilation options.

Not sure how to check this now that I bring it up... `configure` is a specific thing (autotools) and referring to it as "configure options" might be too generic... but a lot of people don't know what autotools even are anymore...

File: content/features/proxy2/install.md
Comment (dormando) [inline content/features/proxy2/install.md]: I'm pretty sure this will confuse people.

The "proxy enabled memcached" is 100% identical to a normal memcached if you don't supply the proxy-related startup options. There's no practical reason to need to keep the binaries separate. A huge part of the convenience of the feature is that it can be both in the same binary.

But on reading this people might think if they build the proxy it'll somehow contaminate the binary.

File: content/features/proxy2/install.md
Comment (dormando) [inline content/features/proxy2/install.md]: `--enable-proxy`

File: content/features/proxy2/install.md
Comment (dormando) [reply content/features/proxy2/install.md]: I think the typical idea here is you can give it a different _installation prefix_, ie: `./configure --enable-proxy --prefix=/usr/local` which will install into `/usr/local/bin` if the user already has a /usr/bin/memcached from the system. This is standard practice if you're on a distro and want a newer binary for some reason.

File: content/features/proxy2/configure.md
Comment (dormando) [inline content/features/proxy2/configure.md]: `The configuration file is a Lua script that extends the standard proxy route library.` Probably best to de-emphasize this at this point. I'd rather call it a data file that uses lua syntax, since the config syntax is using a DSL style.

The moment I say "you write lua to add servers to the thing" people's brains shut off.

File: content/features/proxy2/configure.md
Comment (dormando) [inline content/features/proxy2/configure.md]: one or more. It can actually be called multiple times (for organizational or scripting purposes, I guess)

File: content/features/proxy2/configure.md
Comment (dormando) [inline content/features/proxy2/configure.md]: `verbose(true)` is more lua-y.

File: content/features/proxy2/configure.md
Comment (dormando) [inline content/features/proxy2/configure.md]: `features of the proxy` -> `features of the routing library`. Users might enable these thinking it enables the proxy's own logging (which is confusing enough already..)

File: content/features/proxy2/configure.md
Comment (dormando) [inline content/features/proxy2/configure.md]: routelib is abstracting these calls out (I don't think this even works? heh)

Please compare these routelib snippits with examples from the routelib examples directory: https://github.com/memcached/memcached-proxylibs/blob/main/lib/routelib/examples/basic.lua

you can see the settings block looks a lot more natural.

I'm not sure how to best document this. It might be best to double-document the available settings here.

File: content/features/proxy2/configure.md
Comment (dormando) [inline content/features/proxy2/configure.md]: Do the examples need to be capitalized? I think routelib actually requires lowercase `set_` and people might give everything uppercase names. maybe I could fix that in the library tho...

File: content/features/proxy2/configure.md
Comment (dormando) [inline content/features/proxy2/configure.md:72]: there's also an `options` key for pool-specific options, ie:
https://github.com/memcached/memcached-proxylibs/blob/main/lib/routelib/examples/pooloptions.lua

File: content/features/proxy2/configure.md
Comment (dormando) [inline content/features/proxy2/configure.md]: `route_direct`

File: content/features/proxy2/configure.md
Comment (dormando) [inline content/features/proxy2/configure.md]: `route_direct`

File: content/features/proxy2/configure.md
Comment (dormando) [inline content/features/proxy2/configure.md]: `route_direct`

File: content/features/proxy2/configure.md
Comment (dormando) [inline content/features/proxy2/configure.md]: `set_main_pools`

File: content/features/proxy2/configure.md
Comment (dormando) [inline content/features/proxy2/configure.md]: `children = "set_main_pools"`

see: https://github.com/memcached/memcached-proxylibs/blob/main/lib/routelib/examples/zfailover.lua

File: content/features/proxy2/run.md
Comment (dormando) [inline content/features/proxy2/run.md]: Can this hot link to the how to get routelib section somewhere?

**Review by dormando — CHANGES_REQUESTED** [review]
in `routes{}` section, `cmap` can be a top level thing, ie:

https://github.com/memcached/memcached-proxylibs/blob/main/lib/routelib/examples/commandmap.lua

You can even mix map, cmap, and default:
https://github.com/memcached/memcached-proxylibs/issues/7#issuecomment-2177814872

IE:
- check map for key prefixes for special handling
- if none match, check the cmap for a 'get' override (ie: failover, try pool A first)
- else hit the default, which is allsync for all replicated pools (ie; set commands)

**Comment by dormando** [comment]
I like the vibe of where this is going.

Not sure if you intended for me to fix the smaller things myself, but I noted a bunch of them. Some things are larger issues, ie; missing sections. Or need to make a decision on how to document something best (ie; the `settings` section, which doesn't use `mcp.` function calls)

File: themes/hextra/assets/css/custom.css
Comment (dormando) [reply themes/hextra/assets/css/custom.css]: Think the right place to put this is `static/assets/css/custom.css` which will get deployed to `public/assets/css/custom.css` which the theme should pick up

**Comment by dormando** [comment]
This makes it stand out to me a lot that I need to get routelib bundled in the source code asap... sigh..

File: themes/hextra/assets/css/custom.css
Comment (jmacdotorg) [reply themes/hextra/assets/css/custom.css]: That was where my first attempt led me, but it doesn't seem to work without further steps. I think you're right that this is the cleaner solution but it requires deeper knowledge of Hugo than I currently have, and I didn't see an easy way of getting there that wouldn't have likely involved hours more of research.

Would it be acceptable to keep `custom.css` where it is now, and open an issue to address after the docs are delivered?

File: content/features/proxy2/quickstart.md
Comment (jmacdotorg) [reply content/features/proxy2/quickstart.md:1]: I don't think so. My intent is that the index page should be the feature's conceptual overview, which is separate from the QuickStart. See also #4.

File: themes/hextra/assets/css/custom.css
Comment (dormando) [reply themes/hextra/assets/css/custom.css]: what were the symptoms? it just didn't seem to apply at all? (or what further steps were you expecting?)

File: content/features/proxy2/install.md
Comment (jmacdotorg) [reply content/features/proxy2/install.md:57]: Got it! I've removed all of this suggestion, from this page and others. I wrote extra-cautiously here, but you're the best judge of your users' comfort level and I'm happy to keep things simpler.

File: content/features/proxy2/install.md
Comment (jmacdotorg) [reply content/features/proxy2/install.md]: Fixed.

File: content/features/proxy2/install.md
Comment (jmacdotorg) [reply content/features/proxy2/install.md]: I've renamed the placeholder to OTHER_OPTIONS and rephrased the text to refer to them as "other options for the `configure` command".

File: content/features/proxy2/install.md
Comment (jmacdotorg) [reply content/features/proxy2/install.md]: Got it! I've removed all of this suggestion, from this page and others. I wrote extra-cautiously here, but you're the best judge of your users' comfort level and I'm happy to keep things simpler.

File: content/features/proxy2/install.md
Comment (jmacdotorg) [reply content/features/proxy2/install.md]: Fixed.

File: content/features/proxy2/configure.md
Comment (jmacdotorg) [reply content/features/proxy2/configure.md]: Fixed, and changed the argument to `child = {...}`.

File: content/features/proxy2/configure.md
Comment (jmacdotorg) [reply content/features/proxy2/configure.md]: Fixed, and changed the argument to `child = {...}`.

File: content/features/proxy2/configure.md
Comment (jmacdotorg) [reply content/features/proxy2/configure.md]: How about `main_pool_set`? `set_main_pools` reads like a command to me—not least because `set` is a *very* loaded initial keyword in Memcached-land, if I understand correctly!

File: content/features/proxy2/configure.md
Comment (jmacdotorg) [reply content/features/proxy2/configure.md]: Fixed. I also similarly updated line 358. (And renamed the arguments to `main_pool_set` and `customer_pool_set` for reasons in previous comment.)

File: content/features/proxy2/run.md
Comment (jmacdotorg) [reply content/features/proxy2/run.md]: The "Before you begin" section on this page already states that you need to have downloaded it before any of this will work. That said, there's no harm in clarity, so I've added a little reminder text here; what do you think?

**Comment by jmacdotorg** [comment]
> I like the vibe of where this is going.
> 
> Not sure if you intended for me to fix the smaller things myself, but I noted a bunch of them. Some things are larger issues, ie; missing sections. Or need to make a decision on how to document something best (ie; the `settings` section, which doesn't use `mcp.` function calls)

Let's treat this as a proper code review; point out what's wrong, and I'll revise as needed to get your ultimate approval.

There aren't any *intentionally* missing sections here, so let's chat about what I've overlooked, and I'll make those revisions too.

File: content/features/proxy2/configure.md
Comment (dormando) [reply content/features/proxy2/configure.md]: `child = "thing"` for direct. doesn't use the list syntax since it's singular

File: themes/hextra/assets/css/custom.css
Comment (jmacdotorg) [reply themes/hextra/assets/css/custom.css]: It doesn't apply at all. If it does apply, then the commands on e.g. features/proxy2/install/#download are preceded with `$ ` prompts.

I think what's missing is whatever magic is necessary to make both of the following happen:

1. Put the custom CSS file in whatever collection of static files that Hugo generates.
2. Link to that file from the `<head>` sections of all of the generated HTML pages.

Certainly that's possible—it's probably even a basic Hugo trick. But since I'm coming at all of this rather sideways, I don't know it.

File: content/features/proxy2/configure.md
Comment (dormando) [reply content/features/proxy2/configure.md]: The problem is routelib is doing magic based on the name of the pool. I got stuck with this because I'm planning on making the config JSON compatible in the future so I can't use function call syntax to define pool sets.

It's looking for pools that start with `set_` and then recursing to collect them into a set.

Comment (dormando) [reply content/features/proxy2/configure.md]: I could consider changing that keyword or adding a second one to look for, if you have a good suggestion. it needs to be a prefix though.

Comment (jmacdotorg) [reply content/features/proxy2/configure.md]: I didn't know that! OK, pocketing my writerly opinions in this case, then. I've changed all instances to `set_main_pools` and `set_customer_pools`.

File: content/features/proxy2/configure.md
Comment (jmacdotorg) [reply content/features/proxy2/configure.md]: Done.

File: themes/hextra/assets/css/custom.css
Comment (dormando) [reply themes/hextra/assets/css/custom.css]: I'll take a look at it myself.

It _should_ be trivial though: whether the file is in that hextra directory or the static directory, it should land in the exact same path in the public/ directory once built.

File: content/features/proxy2/install.md
Comment (dormando) [reply content/features/proxy2/install.md]: In my experience the fewer "branches" for people to go down the better.

File: content/features/proxy2/run.md
Comment (dormando) [reply content/features/proxy2/run.md]: Think that's good enough for now. There'll be tons of people who just go to this page directly and get confused at what 'routelib' is, and it's not a link anywhere. So it's good to tell them to go somewhere else.

File: content/features/proxy2/configure.md
Comment (jmacdotorg) [reply content/features/proxy2/configure.md]: Actually, are you saying that the names of named sets *must* start with, literally, the characters `set_`? Because that needs to go in the docs if so.

Comment (dormando) [reply content/features/proxy2/configure.md]: they must start with `set_` if you want it to collect the pools into a set (otherwise it tries to parse it as a normal pool syntax, which obvious won't work).

I could make them use a totally separate main section, ie:
`poolsets{
   foo = { { east = {etc} }, { west = { etc } }
}`

.. but down in the children/child definition I'd still need some kind of special syntax.

from the original docs: https://github.com/memcached/memcached-proxylibs/tree/main/lib/routelib#zones-and-pool-sets
looks like I wrote "special syntax" and then forgot to write the next sentence saying what that syntax is, but provided the example.

**Comment by dormando** [comment]
lmk if I should review again. I see the last code review as still open in GH

**Comment by jmacdotorg** [comment]
> This makes it stand out to me a lot that I need to get routelib bundled in the source code asap... sigh..

Sure thing, will follow up when it's ready for another review.

**Comment by jmacdotorg** [comment]
> in `routes{}` section, `cmap` can be a top level thing, ie:
> 
> https://github.com/memcached/memcached-proxylibs/blob/main/lib/routelib/examples/commandmap.lua
> 
> You can even mix map, cmap, and default: [memcached/memcached-proxylibs#7 (comment)](https://github.com/memcached/memcached-proxylibs/issues/7#issuecomment-2177814872)
> 
> IE:
> 
> * check map for key prefixes for special handling
> * if none match, check the cmap for a 'get' override (ie: failover, try pool A first)
> * else hit the default, which is allsync for all replicated pools (ie; set commands)

I have rewritten the `routes{}` section to add `cmap` and examine that and the other possible sub-sections in greater detail. (https://github.com/memcached/memcached.github.io/pull/3/commits/43bf006c1084a0c6439f6566bbe5de364d0c2dfe)

File: content/features/proxy2/configure.md
Comment (jmacdotorg) [reply content/features/proxy2/configure.md]: > routelib is abstracting these calls out (I don't think this even works? heh)
> 
> Please compare these routelib snippits with examples from the routelib examples directory: https://github.com/memcached/memcached-proxylibs/blob/main/lib/routelib/examples/basic.lua
> 
> you can see the settings block looks a lot more natural.
> 
> I'm not sure how to best document this. It might be best to double-document the available settings here.

I rewrote this section appropriately, and added a new settings reference, in https://github.com/memcached/memcached.github.io/pull/3/commits/e6f61e933771eab9ce6833b6d2a59523ddf2ab47.

File: content/features/proxy2/configure.md
Comment (jmacdotorg) [reply content/features/proxy2/configure.md]: I've added clarity about this with https://github.com/memcached/memcached.github.io/pull/3/commits/7e99460cf4d6f43ffcc669ffe20a8e2a76fe098a

File: content/features/proxy2/configure.md
Comment (jmacdotorg) [reply content/features/proxy2/configure.md]: I've added clarity around set definition with https://github.com/memcached/memcached.github.io/pull/3/commits/7e99460cf4d6f43ffcc669ffe20a8e2a76fe098a, including moving it into its own sub-section and making the naming rules more clear.

File: content/features/proxy2/configure.md
Comment (jmacdotorg) [reply content/features/proxy2/configure.md:72]: Added with https://github.com/memcached/memcached.github.io/pull/3/commits/5698658af884c6fed376fb937b18f0b78a9ae1cb.

File: content/features/proxy2/configure.md
Comment (jmacdotorg) [reply content/features/proxy2/configure.md]: Rewrote the section to say this:

> Before you can run the proxy, you need to create a configuration that lists the pools of Memcached servers that you want to use as the proxy's back end, and describes the routing trees that you want the proxy to use.
> 
> The configuration extends the standard proxy route library. It uses Lua syntax, but you don't need to know how to program in Lua to create and modify the configuration file.
> 
> While advanced users can modify the route library itself, the techniques described on this page focus on using the route library as-is, which covers the majority of Memcached proxy use-cases.

File: content/features/proxy2/configure.md
Comment (jmacdotorg) [reply content/features/proxy2/configure.md]: Done.

File: content/features/proxy2/configure.md
Comment (jmacdotorg) [reply content/features/proxy2/configure.md]: Done, here and in two other places.

File: content/features/proxy2/configure.md
Comment (jmacdotorg) [reply content/features/proxy2/configure.md]: Changed to:

>         main = route_direct {
>             child = "main_pool",
>         },

File: content/features/proxy2/configure.md
Comment (jmacdotorg) [reply content/features/proxy2/configure.md]: Done, and added a clarifying paragraph to the start of the pools{} section.

**Comment by dormando** [comment]
I don't have any major complaints on a first-pass reading here, thanks much!

I'll re-read it again later today and see if I can get more nit-picky

File: content/features/proxy2/configure.md
Comment (dormando) [inline content/features/proxy2/configure.md:406]: Would it be bad to add a second example to set usage here?

There's a "non zoned" syntax where you just use an array instead of a map and that seems to be popular among the simple use cases: https://github.com/memcached/memcached-proxylibs/issues/11

I've fielded a few of these now

**Comment by dormando** [comment]
Just the one standalone comment here. I'll check the css thing today.

I think any minor factual issues I can fix myself when I'm working on it.

File: themes/hextra/assets/css/custom.css
Comment (dormando) [reply themes/hextra/assets/css/custom.css]: @jmacdotorg I got this working. steps:

- revert the change from themes/hextra/etc
- put your custom.css file in `assets/css/custom.css` _from the top level of the repository_
ie: the same directory that hugo.yaml is in, make an assets dir, then a css subdir, then put the custom.css in there. keeping the theme/hextra/etc/custom.css blank as per default.

I've been testing locally and editing that assets/css/custom.css file makes the $ change come and go.

File: content/features/proxy2/configure.md
Comment (jmacdotorg) [reply content/features/proxy2/configure.md:406]: Good idea, but I don't think this is the place to put that example; this section is specifically about setting up zones.

There is an earlier section in the document about non-zoned sets; see the heading "Define sets of pools". I've just now put a new example into it. Have a look! (https://github.com/memcached/memcached.github.io/pull/3/commits/c5f447fbf3ed8820c3051ddc3fb71c757af63dbc)

File: themes/hextra/assets/css/custom.css
Comment (jmacdotorg) [reply themes/hextra/assets/css/custom.css]: Excellent work! This works for me too, and I've updated this pull request appropriately. (https://github.com/memcached/memcached.github.io/pull/3/commits/10c845ce41c181ed1b991b8ff409354b4357e3f9)

**Comment by dormando** [comment]
Squashed it down to one commit to minimize fiddling and merged, thanks!

**Comment by dormando** [comment]
wow okay. I totally fucked up that merge somehow. instead of rebasing it looks like it just nuked half the branch...

will try it again.

**Comment by dormando** [comment]
okay fixed now. I probably fat-fingered an r for an f during the rebase

**Pull request closed without merging**
